### PR TITLE
Don't use system() to ZIP files in auxiliary/client/smtp/emailer 

### LINKED ISSUE #11286: auxiliary/client/smtp/emailer uses system() to zip files
The Generic Emailer (SMTP) module uses `system()` rather than `Rex::Zip::Archive.new` to compress attachment files.

This could lead to command injection in the unlikely event that the user specifies a malicious YAML file in the `YAML_CONFIG` option prior to executing the module.

It's also worth noting that due to this approach, the module will not work on systems which do not have a `zip` utility in `$PATH`.

The relevant source is shown below:

```ruby
    attachment_file      = yamlconf['attachment_file']
# [...]
      if zip_payload
        zip_file = attachment_file.sub(/\.\w+$/, '.zip')
        system("zip -r #{zip_file} #{attachment_file}> /dev/null 2>&1");
        attachment_file      = zip_file
        attachment_file_type = 'application/zip'
      else
        attachment_file_type = 'application/exe'
      end
```

$ ./msfconsole -x "use auxiliary/client/smtp/emailer; set yaml_config data/emailer_config.yaml; run"

```
                                                  
[%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%]
[%%%%%%%%%%%%%%%%%%%%%%.,-:;//;:=,%%%%%%%%%%%%%%%%%%%%%%%%]
[%%%%%%%%%%%%%%%%%%.%:H@@@MM@M#H/.,+%;,%%%%%%%%%%%%%%%%%%%]
[%%%%%%%%%%%%%%%,/X+%+M@@M@MM%=,-%HMMM@X/,%%%%%%%%%%%%%%%%]
[%%%%%%%%%%%%%-+@MM;%$M@@MH+-,;XMMMM@MMMM@+-%%%%%%%%%%%%%%]
[%%%%%%%%%%%%;@M@@M-%XM@X;.%-+XXXXXHHH@M@M#@/.%%%%%%%%%%%%]
[%%%%%%%%%%,%MM@@MH%,@%=%%%%%%%%%%%%%.---=-=:=,.%%%%%%%%%%]
[%%%%%%%%%%=@#@@@MX.,%%%%%%%%%%%%%%%%-%HX$$%%%:;%%%%%%%%%%]
[%%%%%%%%%=-./@M@M$%%%%%%%%%%%%%%%%%%%.;@MMMM@MM:%%%%%%%%%]
[%%%%%%%%%X@/%-$MM/%%%%%%%%%%%%%%%%%%%%.%+MM@@@M$%%%%%%%%%]
[%%%%%%%%,@M@H:%:@:%%%%%%%%%%%%%%%%%%%%.%=X#@@@@-%%%%%%%%%]
[%%%%%%%%,@@@MMX,%.%%%%%%%%%%%%%%%%%%%%/H-%;@M@M=%%%%%%%%%]
[%%%%%%%%.H@@@@M@+,%%%%%%%%%%%%%%%%%%%%%MM+..%#$.%%%%%%%%%]
[%%%%%%%%%/MMMM@MMH/.%%%%%%%%%%%%%%%%%%XM@MH;%=;%%%%%%%%%%]
[%%%%%%%%%%/%+%$XHH@$=%%%%%%%%%%%%%%,%.H@@@@MX,%%%%%%%%%%%]
[%%%%%%%%%%%.=--------.%%%%%%%%%%%-%H.,@@@@@MX,%%%%%%%%%%%]
[%%%%%%%%%%%.%MM@@@HHHXX$$$%+-%.:$MMX%=M@@MM%.%%%%%%%%%%%%]
[%%%%%%%%%%%%%=XMMM@MM@MM#H;,-+HMM@M+%/MMMX=%%%%%%%%%%%%%%]
[%%%%%%%%%%%%%%%=%@M@M#@$-.=$@MM@@@M;%%M%=%%%%%%%%%%%%%%%%]
[%%%%%%%%%%%%%%%%%,:+$+-,/H#MMMMMMM@=%=,%%%%%%%%%%%%%%%%%%]
[%%%%%%%%%%%%%%%%%%%%%%%=++%%%%+/:-.%%%%%%%%%%%%%%%%%%%%%%]
[%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%]


       =[ metasploit v4.15.0-dev-d02aa6f                  ]
+ -- --=[ 1680 exploits - 959 auxiliary - 294 post        ]
+ -- --=[ 486 payloads - 40 encoders - 9 nops             ]
+ -- --=[ Free Metasploit Pro trial: http://r-7.co/trymsp ]

yaml_config => data/emailer_config.yaml
[*] 127.0.0.1:25 - Creating payload...
[*] 127.0.0.1:25 - Writing payload to /tmp/`nc -lvp 1337 -e bash` #
nc: listening on :: 1337 ...
nc: listening on 0.0.0.0 1337 ...
^C[-] 127.0.0.1:25 - Auxiliary interrupted by the console user
[*] Auxiliary module execution completed
```

$ cat data/emailer_config.yaml

```yaml
# YAML:1.0
# Configuration file for emailer.rb module
# Taken from Jabra spl0it.org
#
# Emails CSV File (Fname Lname,email)
to: /dev/null
# Email is sent from this address
from: root@localhost
# Subject
subject: "[SECURITY-ALERT] Critical Windows Vulnerability"
# Type ( text or text/html )
type: text/html
# Msg body file
msg_file: /dev/null
# Number of seconds to wait before next email
wait: 5
# Prepend the first name to the email body
add_name: true
# Add custom signature from file
sig: true
# Signature file
sig_file: /dev/null

#### Attachment Information ####
# Add an email attachment (File exploit anyone?) 
attachment: true
# Path to file attachment
attachment_file: test.jpg
# Name of file attachment
attachment_file_name: msf.jpg
# Type of attachment
attachment_file_type: image/jpeg

#### Metasploit/Payload Creation ####
# create a metasploit payload
make_payload: true
# zip the payload 
zip_payload: true
# metasploit server ip
msf_ip: 127.0.0.1
# metasploit server port
msf_port: 443
# metasploit payload
msf_payload: windows/meterpreter/reverse_tcp
# metasploit payload
msf_filename: '`nc -lvp 1337 -e bash` #'
# metasploit location
msf_location: /dev/null
# change the extension
msf_change_ext: false
# new extension
msf_payload_ext: vxe
```

### PR DESCRIPTION
**Pull request by wvu**
```
msf5 auxiliary(client/smtp/emailer) > run

[*] 127.0.0.1:25 - Creating payload...
[*] 127.0.0.1:25 - Writing payload to /var/folders/mf/ddclmcss6z9df5nj3fzwn78wjsnnkf/T/`nc -lvp 1337 -e bash` #
[*] 127.0.0.1:25 - Zipping payload to /var/folders/mf/ddclmcss6z9df5nj3fzwn78wjsnnkf/T/`nc -lvp 1337 -e bash` #
[*] 127.0.0.1:25 - Email sent..
[*] Auxiliary module execution completed
msf5 auxiliary(client/smtp/emailer) > unzip -l '/var/folders/mf/ddclmcss6z9df5nj3fzwn78wjsnnkf/T/`nc -lvp 1337 -e bash` #'
[*] exec: unzip -l '/var/folders/mf/ddclmcss6z9df5nj3fzwn78wjsnnkf/T/`nc -lvp 1337 -e bash` #'

Archive:  /var/folders/mf/ddclmcss6z9df5nj3fzwn78wjsnnkf/T/`nc -lvp 1337 -e bash` #
  Length      Date    Time    Name
---------  ---------- -----   ----
    73802  01-24-2019 13:53   `nc -lvp 1337 -e bash` #
---------                     -------
    73802                     1 file
msf5 auxiliary(client/smtp/emailer) >
```

Fixes #11286.

### CODE REVIEW

**Comment by bcoles** [comment]
Seems reasonable.

It looks like there may be another bug in the existing implementation.

I would have expected a `.zip` extension, based on `attachment_file.sub(/\.\w+$/, '.zip')`, in the output above.

```
[*] 127.0.0.1:25 - Writing payload to /var/folders/mf/ddclmcss6z9df5nj3fzwn78wjsnnkf/T/`nc -lvp 1337 -e bash` #
[*] 127.0.0.1:25 - Zipping payload to /var/folders/mf/ddclmcss6z9df5nj3fzwn78wjsnnkf/T/`nc -lvp 1337 -e bash` #
```

```
2.3.0 :001 > attachment_file = 'asdf'
 => "asdf" 
2.3.0 :002 > zip_file = attachment_file.sub(/\.\w+$/, '.zip')
 => "asdf" 
2.3.0 :003 > attachment_file = 'asdf.txt'
 => "asdf.txt" 
2.3.0 :004 > zip_file = attachment_file.sub(/\.\w+$/, '.zip')
 => "asdf.zip" 
2.3.0 :005 > 
```

Would that not be bad, in that it replaces the existing file with a compressed file (but only if the existing file did not have a file extension) ?

File: modules/auxiliary/client/smtp/emailer.rb [hunk 1]
```diff
@@ -161,7 +159,8 @@ def run
 
       if zip_payload
         zip_file = attachment_file.sub(/\.\w+$/, '.zip')
```
Comment (bcoles) [inline modules/auxiliary/client/smtp/emailer.rb]: Is this a reasonable suggestion? (as per https://github.com/rapid7/metasploit-framework/pull/11310#issuecomment-457337672)

```suggestion
        zip_file = "#{attachment_file}.zip"
```

**Review by bcoles — COMMENTED** [review]
LGTM

**Review by bcoles — APPROVED** [review]

**Comment by wchen-r7** [comment]
# Release Notes

This allows the auxiliary/client/smtp/emailer module to use the MSF API to zip instead of relying on the system command.

**Pull request merged**